keyfunc/get.go

```diff
@@ -0,0 +1,223 @@
+package keyfunc
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+)
+
+var (
+
+	// defaultRefreshTimeout is the default duration for the context used to create the HTTP request for a refresh of
+	// the JWKs.
+	defaultRefreshTimeout = time.Minute
+)
+
+// Get loads the JWKs at the given URL.
+func Get(jwksURL string, options ...Options) (jwks *JWKs, err error) {
+
+	// Create the JWKs.
+	jwks = &JWKs{
+		jwksURL: jwksURL,
+	}
+
+	// Apply the options to the JWKs.
+	for _, opts := range options {
+		applyOptions(jwks, opts)
+	}
+
+	// Apply some defaults if options were not provided.
+	if jwks.client == nil {
+		jwks.client = http.DefaultClient
+	}
+	if jwks.refreshTimeout == nil {
+		jwks.refreshTimeout = &defaultRefreshTimeout
+	}
+
+	// Get the keys for the JWKs.
+	if err = jwks.refresh(); err != nil {
+		return nil, err
+	}
+
+	// Check to see if a background refresh of the JWKs should happen.
+	if jwks.refreshInterval != nil || jwks.refreshUnknownKID {
+
+		// Attach a context used to end the background goroutine.
+		jwks.ctx, jwks.cancel = context.WithCancel(context.Background())
+
+		// Create a channel that will accept requests to refresh the JWKs.
+		jwks.refreshRequests = make(chan context.CancelFunc, 1)
+
+		// Start the background goroutine for data refresh.
+		go jwks.backgroundRefresh()
+	}
+
+	return jwks, nil
+}
+
+// backgroundRefresh is meant to be a separate goroutine that will update the keys in a JWKs over a given interval of
+// time.
+func (j *JWKs) backgroundRefresh() {
+
+	// Create some rate limiting assets.
+	var lastRefresh time.Time
+	var queueOnce sync.Once
+	var refreshMux sync.Mutex
+	if j.refreshRateLimit != nil {
+		lastRefresh = time.Now().Add(-*j.refreshRateLimit)
+	}
+
+	// Create a channel that will never send anything unless there is a refresh interval.
+	refreshInterval := make(<-chan time.Time)
```

Comment:
could this work with a time.Ticker?

Reply:
If specifying the `RefreshInterval` of the `keyfunc.Options` had a non-zero default, yes.

Since it can be omitted with the expectation of no refresh interval, a `time.Ticker` cannot be created because it requires a `time.Duration` greater than zero.

If it were turned into a `time.Ticker` now, then the below `select` statement case would need to be guarded by an `if refreshInterval != nil` statement, which would require two `select` statements, once for each case.